presto-main/src/main/java/com/facebook/presto/sql/analyzer/MaterializedViewInformationExtractor.java

```diff
@@ -0,0 +1,174 @@
+            throw new SemanticException(NOT_SUPPORTED, node, "Having clause is not supported in query optimizer");
+        }
+        super.visitQuerySpecification(node, context);
+        return null;
+    }
+
+    protected Void visitSelect(Select node, Void context)
+    {
+        super.visitSelect(node, context);
+        materializedViewInfo.setDistinct(node.isDistinct());
+        return null;
+    }
+
+    @Override
+    protected Void visitRelation(Relation node, Void context)
+    {
+        if (node instanceof Table) {
+            if (materializedViewInfo.getBaseTable().isPresent()) {
+                throw new SemanticException(NOT_SUPPORTED, node, "Only support single table rewrite in query optimizer");
+            }
+            materializedViewInfo.setBaseTable(Optional.of(node));
+            return null;
+        }
+        throw new SemanticException(NOT_SUPPORTED, node, "Relation other than Table is not supported in query optimizer");
+    }
+
+    @Override
+    protected Void visitSingleColumn(SingleColumn node, Void context)
+    {
+        Expression baseColumnName = node.getExpression();
+        materializedViewInfo.addBaseToViewColumn(baseColumnName, node.getAlias().orElse(new Identifier(baseColumnName.toString())));
+        return null;
+    }
+
+    @Override
+    protected Void visitAllColumns(AllColumns node, Void context)
+    {
+        throw new SemanticException(NOT_SUPPORTED, node, "All columns materialized view is not supported in query optimizer");
+    }
+
+    @Override
+    protected Void visitGroupBy(GroupBy node, Void context)
+    {
+        materializedViewInfo.setGroupBy(Optional.of(ImmutableSet.copyOf(node.getGroupingElements())));
+        return null;
+    }
+
+    @Override
+    protected Void visitLogicalBinaryExpression(LogicalBinaryExpression node, Void context)
+    {
+        materializedViewInfo.setWhereClause(Optional.of(node));
+        return null;
+    }
+
+    @Override
+    protected Void visitComparisonExpression(ComparisonExpression node, Void context)
+    {
+        materializedViewInfo.setWhereClause(Optional.of(node));
+        return null;
+    }
+
+    public MaterializedViewInfo getMaterializedViewInfo()
+    {
+        return materializedViewInfo;
+    }
+
+    public static final class MaterializedViewInfo
+    {
+        private final Map<Expression, Identifier> baseToViewColumnMap = new HashMap<>();
+        private Optional<Relation> baseTable = Optional.empty();
+        private Optional<Expression> whereClause = Optional.empty();
+        private Optional<ImmutableSet<GroupingElement>> groupBy = Optional.empty();
+        private boolean isDistinct;
+
+        private void addBaseToViewColumn(Expression key, Identifier value)
+        {
+            baseToViewColumnMap.put(key, value);
+        }
+
+        private void setGroupBy(Optional<ImmutableSet<GroupingElement>> groupBy)
+        {
+            this.groupBy = groupBy;
```

Comment:
I guess we can have `checkState(!this.groupBy.isPresent());` before this? Is this also true for all other setters?

Reply:
@highker This is true for all setters, I'll add a checkstate for each of them.